stonesoup/mixturereducer/gaussianmixture.py

```diff
@@ -0,0 +1,210 @@
+        if isinstance(component_1, TaggedWeightedGaussianState):
+            merged_component = TaggedWeightedGaussianState(
+                state_vector=merged_mean,
+                covar=merged_covar,
+                weight=merged_weight,
+                tag=component_1.tag,
+                timestamp=component_1.timestamp
+            )
+        elif isinstance(component_1, WeightedGaussianState):
+            merged_component = WeightedGaussianState(
+                state_vector=merged_mean,
+                covar=merged_covar,
+                weight=merged_weight,
+                timestamp=component_1.timestamp
+            )
+
+        return merged_component
+
+    def merge(self, components_list):
+        """
+        Merging is the act of combining similar components in the mixture
+        that fall with a distance threshold :attr:`merge_threshold` into
+        a single component.
+
+        Parameters
+        ----------
+        components_list : :class:`~.list`
+            Components of the Gaussian Mixture to be merged
+
+        Returns
+        -------
+        :class:`~.list`
+            Merged components
+
+        """
+        # Sort components by weight
+        remaining_components = sorted(
+            components_list, key=attrgetter('weight'))
+
+        merged_components = []
+        final_merged_components = []
+        while remaining_components:
+            # Get highest weighted component
+            best_component = remaining_components.pop()
+            # Check for similar components
+            # (modifying list in loop, so copy used)
+            for component in remaining_components.copy():
+                # Calculate distance between component and best component
+                distance = dist.mahalanobis(
+                    best_component.mean, component.mean, best_component.covar)
+                # Merge if similar
+                if distance < self.merge_threshold:
+                    remaining_components.remove(component)
+                    best_component = self.merge_components(
+                        best_component, component
+                    )
+            # Add potentially merged component to new mixture
+            merged_components.append(best_component)
+        if all(isinstance(component, TaggedWeightedGaussianState)
+               for component in merged_components):
+            # Check for duplicate tags
+            components_tags = set(component.tag for component in merged_components)
+            if len(components_tags) != len(merged_components):
+                # There are duplicatze tags so assign
+                # new tags to the lower weighted shared ones
+                for shared_tag in components_tags:
+                    shared_indices = [i for i, x in enumerate(merged_components)
+                                      if x.tag == shared_tag]
+                    shared_components = [merged_components[i] for i in shared_indices]
```

Comment:
Don't need indices here:
```python
shared_components = [component for component in merged_components
                     if component.tag == shared_tag]
```
Could also sort at the same time, to avoid iterating twice, and assuming not many components share the same tag this will probably work out faster. At this point, you could then also remove the need for next if statement. (not that speed is real issue, but just less branching code is nicer).
```python
shared_components = sorted(
    (component for component in merged_components
     if component.tag == shared_tag),
    key=attrgetter('weight'))
```